enhancements/machine-config/mco-network-configuration.md

```diff
@@ -0,0 +1,292 @@
+---
+title: mco-network-configuration
+authors:
+  - "@bcrochet"
+reviewers:
+  - "@cwalters"
+approvers:
+  - TBD
+creation-date: 2020-06-25
+last-updated: 2020-06-25
+status: provisional
+---
+
+# Extend MCO to support declarative host network configuration [baremetal]
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The ability to configure additional host networking parameters is paramount on some platforms, especially baremetal. Baremetal nodes usually have many interfaces that need to be dynamically configured to access other, critical datacenter resources. Doing so with MachineConfig would incur reboots that are expensive in time on the baremetal platform. The ability to configure host networking parameters without rebooting is desirable. It is also desirable to rollback automatically on failure, as opposed to failing on reboot.
+
+## Motivation
+
+With hybrid clouds, node-networking setup is becoming even more challenging.
+Different payloads have different networking requirements, and not everything
+can be satisfied as overlays on top of the main interface of the node (e.g.
+SR-IOV, L2, other L2).
+
+The [Container Network Interface](https://github.com/containernetworking/cni) (CNI) standard enables different solutions for connecting networks on the node with pods. Some of them are [part of the standard](https://github.com/containernetworking/plugins), and there are others that extend support for [Open vSwitch bridges](https://github.com/kubevirt/ovs-cni), [SR-IOV](https://github.com/hustcat/sriov-cni), and more...
+
+However, in all of these cases, the node must have the networks setup before the pod is scheduled. Setting up the networks in a dynamic and heterogenous cluster, with dynamic networking requirements, is a challenge by itself - and this is what this project is addressing.
+
+The node network configuration shall remain in the MachineConfigOperator workflow, but be declarative.
+
+### Goals
+
+* Enable an admin to modify node networking configuration without rebooting (Day 2)
+* Enable cluster administrators to modify node networking after cluster installation (day 2).
+* Integration with MachineConfig node configuration for consistency, familiarity and roll out functionality.
+* Rebootless configuration without deferring errors to a later reboot
+* Rolling back cluster-breaking configuration mistakes.
+
+### Non-Goals
+
+* Replace SRIOV operator
+* Configure and control primary interface on Day 1, including bonds and VLANs.
+  * Part of the effort for that: [Add MCO Flattened Ignition proposal](https://github.com/openshift/enhancements/pull/467)
```

Comment:
@bcrochet thanks for adding the notes here, one minor follow-up, I think the proposal is not going to work in day 2 for primary node interface, agree?

Reply:
I would assume not. But don't quote me on that.

Reply:
Would it make sense to update this statement to `Configure and control primary interface on both Day 1 and Day 2` as the non-goal?

Reply:
Is there any protection against trying to use this to adjust primary interface configuration if it's not intended to be used that way?

Reply:
I don't think there is a protection mechanism in current proposal (this is also true for additional host interfaces if multiple components try to configure the same device), my understanding of the flow here is it will fail in the case that network configuration is wrong, deleting the NNCP + rebooting node(s) would erase the ephemeral network configuration and recover.

Reply:
In the failure scenario, it wouldn't even require a reboot. But to answer @russellb , no, there is no protection other than that the admin would have to affirmatively try to do it.